source/common/event/timer_impl.h

```diff
@@ -16,7 +16,33 @@ namespace Event {
  */
 class TimerUtils {
 public:
-  static void millisecondsToTimeval(const std::chrono::milliseconds& d, timeval& tv);
+  /**
+   * Intended for consumption by enable(HR)Timer, this method is templated method to avoid implicit
+   * duration conversions for its input arguments. This lets us have an opportunity to check bounds
+   * before doing any conversions. When the passed in duration exceeds UINT32_T max days, the output
+   * will be clipped to yield INT32_MAX days worth of seconds and 0 microseconds for the output
+   * argument. ASSERTS on negative duration input.
+   * @tparam Duration std::chrono duration type, e.g. seconds, milliseconds, ...
+   * @param d duration value
+   * @param tv output parameter that will be updated
+   */
+  template <typename Duration> static void durationToTimeval(const Duration& d, timeval& tv) {
+    if (d.count() < 0) {
+      throw EnvoyException(
+          fmt::format("Negative duration passed to durationToTimeval(): {}", d.count()));
+    };
+    constexpr int64_t clip_to = INT32_MAX; // 136.102208 years
```

Comment:
I certainly think this is more than enough. Out of curiosity though, why 32 bits? I think timeval uses time_t for seconds, which is 64 bit?